apps/meteor/client/components/GenericUpsellModal.tsx

```diff
@@ -0,0 +1,102 @@
+import type { Icon } from '@rocket.chat/fuselage';
+import { Box, Button, Modal } from '@rocket.chat/fuselage';
+import { useRouter, useSetModal, useSetting, useTranslation } from '@rocket.chat/ui-contexts';
+import type { ReactNode, ReactElement, ComponentProps } from 'react';
+import React, { useCallback } from 'react';
+
+import ModalBackdrop from './modal/ModalBackdrop';
+
+type UpsellModalProps = {
+	children?: ReactNode;
+	tagline?: ReactNode;
+	cancelText?: ReactNode;
+	confirmText?: ReactNode;
+	title: string | ReactElement;
+	subtitle?: string | ReactElement;
+	description?: string | ReactElement;
+	icon?: ComponentProps<typeof Icon>['name'];
+	img: ComponentProps<typeof Modal.HeroImage>['src'];
+	onCancel?: () => void;
+	onClose?: () => void;
+	onConfirm?: () => void;
+	onCloseEffect?: () => void;
+};
+
+const GenericUpsellModal = ({
+	tagline,
+	title,
+	subtitle,
+	img,
+	cancelText,
+	confirmText,
+	icon,
+	description,
+	onCancel,
+	onConfirm,
+	onClose = onCancel,
+	onCloseEffect,
+}: UpsellModalProps) => {
+	const t = useTranslation();
+	const cloudWorkspaceHadTrial = Boolean(useSetting('Cloud_Workspace_Had_Trial'));
+
+	const router = useRouter();
+	const setModal = useSetModal();
+
+	const handleModalClose = useCallback(() => {
+		setModal(null);
+		if (onCloseEffect) {
+			onCloseEffect();
+		}
+	}, [onCloseEffect, setModal]);
+
+	const handleConfirmModal = useCallback(() => {
+		handleModalClose();
+		router.navigate({
+			pathname: '/admin/upgrade/go-fully-featured-registered',
+		});
+	}, [handleModalClose, router]);
+
+	const talkToSales = 'https://go.rocket.chat/i/contact-sales';
+	const handleCancelModal = useCallback(() => {
+		handleModalClose();
+		window.open(talkToSales, '_blank');
+	}, [handleModalClose]);
+
+	return (
+		<ModalBackdrop onDismiss={onClose ?? handleModalClose}>
```

Comment:
Instead of rendering an additional backdrop to handle access to a dismissing event, we can rely on a simple `useEffect` clean up function to invoke the closing callbacks, _but there is a catch_:

```ts
const onCloseRef = useRef(onClose ?? handleModalClose); // store the callback in a reference
onCloseRef.current = onClose ?? handleModalClose; // updated the callback on each render

useEffect(() => {
  // empty effect, no need to do something on MOUNT
  
  return () => {
  // this will run only on UNMOUNT
    const onClose = onCloseRef.current;
    onClose?.();
  }
}, []); // the empty array for dependencies enables the effect and its clean up to run only when the component MOUNTS and UNMOUNTS
```